field/TinyLang/Field/Evaluator.hs

```diff
@@ -204,72 +265,97 @@ denoteUniConst (UniConst Vector v) = unAField . fromInteger . packPositiveAsc $
 denoteSomeUniConst :: Field f => SomeUniConst f -> f
 denoteSomeUniConst = forget denoteUniConst
 
-denoteExpr
-    :: (MonadEvalError f m, Eq f, Field f, AsInteger f)
-    => Env (SomeUniConst f) -> Expr f a -> m f
-denoteExpr env = fmap denoteUniConst . evalExprUni env
-
-normStatement
-    :: (MonadEvalError f m, Eq f, Field f, AsInteger f)
-    => Env (SomeUniConst f) -> Statement f -> Expr f a -> m (Expr f a)
-normStatement env (ELet (UniVar uni var) def) rest = do
-    defN <- normExpr env def
+denoteExpr :: (Monad m, Eq f, Field f, AsInteger f)
+    => Expr f a -> EvalT f m f
+denoteExpr = fmap denoteUniConst . evalExprUni
+
+-- | A recursive normalizer for programs.
+normProgram :: (Monad m, Eq f, Field f, AsInteger f)
+    => Program f -> EvalT f m (Program f)
+normProgram (Program stmts) = normStatements stmts (pure . Program)
+
+-- | A recursive normalizer for statements
+normStatements :: (Monad m, Eq f, Field f, AsInteger f)
+    => Statements f -> (Statements f -> EvalT f m r) -> EvalT f m r
+normStatements (Statements []) kont       = kont $ Statements []
+normStatements (Statements (s:rest)) kont = normStatement s $ \case
+          Nothing -> normStatements (Statements rest) kont
+          Just sN -> normStatements (Statements rest) $ \ (Statements restN) ->
+              kont $ Statements $ sN : restN
+
+-- | Normalise single statement
+normStatement :: (Monad m, Eq f, Field f, AsInteger f)
+    => Statement f -> (Maybe (Statement f) -> EvalT f m r) -> EvalT f m r
+normStatement (ELet uniVar@(UniVar _ var) def) kont = do
+    defN <- normExpr def
     case defN of
-        EConst uniConst -> normExpr (insertVar var (Some uniConst) env) rest
-        _             -> EStatement (ELet (UniVar uni var) defN) <$> normExpr env rest
-normStatement env (EAssert expr) rest = do
-    exprN <- normExpr env expr
-    EStatement (EAssert exprN) <$> normExpr env rest
-
--- | A recursive normalizer for expressions.
-normExpr
-    :: (MonadEvalError f m, Eq f, Field f, AsInteger f)
-    => Env (SomeUniConst f) -> Expr f a -> m (Expr f a)
-normExpr _   (EConst uniConst) = EConst <$> normUniConst uniConst
-normExpr env expr@(EVar uniVar@(UniVar uni var)) =
+        EConst uniConst ->
+            withVar var (Some uniConst) $ kont Nothing
+        _ ->
+            kont . Just $ ELet uniVar defN
+normStatement (EAssert expr) kont = do
+    exprN <- normExpr expr
+    case exprN of
+      EConst (UniConst _ False) -> throwError $ AssertionFailedEvalError expr
+      EConst (UniConst _ True) -> kont Nothing
+      _ -> kont . Just $ EAssert exprN
+normStatement (EFor uniVar@(UniVar _ var) start end body) kont =
+    normStatements body $ \bodyN ->
+        let actualEnd = max start end in
+        withVar var (Some $ fromIntegral actualEnd) $ case bodyN of
+            Statements [] -> kont Nothing
+            _             -> kont . Just $ EFor uniVar start end bodyN
```

Comment:
Technically speaking, normalizing a loop amounts to unrolling it. I don't know if we want to unroll the loop here, but please add a comment that we could.

Do you need this big-step normalization for something (apart from tests in `Textual.hs`)? I think I was going to use it for the shrinker, but now that we have a small-step normalizer there, we may not need the big-step one. I don't use normalization in the compiler.

Let's keep the code in this PR and remove it later, if we decide we don't need it.